Source/WebCore/platform/audio/gstreamer/AudioDecoderGStreamer.cpp

```diff
@@ -334,16 +335,14 @@ void GStreamerInternalAudioDecoder::decode(std::span<const uint8_t> frameData, b
             m_harness->processOutputSamples();
         else
             m_outputCallback(makeUnexpected("Decode error"_s));
-
-        callback({ });
     });
+    return AudioDecoder::DecodePromise::createAndResolve();
```

Comment:
here I'm not 100% on this, we are resolving the promise concurrently to calling `processOutputSamples()` this is a different code flow than before.

Could we resolve the promise there instead?

Reply:
Hm yes, this might be racy now. I'm taking a look right now, please don't merge this yet.

Reply:
@youennf can you apply this please? It makes the audio decoder behavior consistent with the video decoder.

```diff
diff --git a/Source/WebCore/platform/audio/gstreamer/AudioDecoderGStreamer.cpp b/Source/WebCore/platform/audio/gstreamer/AudioDecoderGStreamer.cpp
index a34f180817ab..aa91f200c333 100644
--- a/Source/WebCore/platform/audio/gstreamer/AudioDecoderGStreamer.cpp
+++ b/Source/WebCore/platform/audio/gstreamer/AudioDecoderGStreamer.cpp
@@ -325,15 +325,15 @@ Ref<AudioDecoder::DecodePromise> GStreamerInternalAudioDecoder::decode(std::span
         GST_BUFFER_DURATION(encodedData.get()) = *duration;
 
     auto result = m_harness->pushSample(adoptGRef(gst_sample_new(encodedData.get(), m_inputCaps.get(), &segment, nullptr)));
+    if (result)
+        m_harness->processOutputSamples();
     m_postTaskCallback([weakThis = WeakPtr { *this }, this, result]() mutable {
         if (!weakThis)
             return;
         if (m_isClosed)
             return;
 
-        if (result)
-            m_harness->processOutputSamples();
-        else
+        if (!result)
             m_outputCallback(makeUnexpected("Decode error"_s));
     });
     return AudioDecoder::DecodePromise::createAndResolve();
```

Reply:
This raciness would be observable with GC issues.
There is no race if `m_harness->processOutputSamples()` calls synchronously m_outputCallback.
That said, I think this should not be more racy with this patch than before.

Reply:
The `processOutputSamples()` will synchronously call the callback passed to the `GStreamerElementHarness::create()` function.

Reply:
so `m_outputCallback` ends-up being called, but from `m_postTaskCallback()`.

Reply:
Yep, I'll do the follow-up of removing `m_outputCallback` today or tomorrow.
This will indeed make the code clearer.

Reply:
I'll make the gstreamer audio change, thanks for checking it.
I am surprised this audio decoding issue was not hit by GTK/WPE bots on WPT audio tests.

Reply:
TBH I haven't noticed a change in tests behavior with this change and without. At least when running the tests locally. Let's see how the bots behave after landing this!

Reply:
@philn  with your change we should call m_postTaskCallback only if `!result` is true ? so that there's no need to capture `result`

Reply:
Right... but anyway iiuc the postTaskCallback stuff is going away soon?